sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/TableHeaderReferenceCheck.java

```diff
@@ -0,0 +1,398 @@
+        columns.add(column);
+      }
+    }
+
+    public void newRow() {
+      rows.add(new RowBuilder());
+      currentRow = rows.get(rows.indexOf(currentRow) + 1);
+    }
+
+    public void newColumn(Table.Column column) {
+      if (!rows.isEmpty()) {
+        TagNode node = column.node();
+        if (hasColSpan(node) && hasRowSpan(node)) {
+          newColumnWithRowColSpan(column);
+        } else if (hasColSpan(node)) {
+          newColumnWithColSpan(column);
+        } else if (hasRowSpan(node)) {
+          newColumnWithRowSpan(column);
+        } else {
+          newColumnWithoutSpan(column);
+        }
+      }
+    }
+
+    private void newColumnWithoutSpan(Table.Column column) {
+      int columnIndex = currentRow.indexOfVacantColumn();
+      if (columnIndex > -1) {
+        currentRow.set(columnIndex, column);
+      } else {
+        currentRow.add(column);
+      }
+    }
+
+    private void newColumnWithColSpan(Table.Column column) {
+      int colspan = getColSpan(column.node());
+      while (colspan > 0) {
+        int columnIndex = currentRow.indexOfVacantColumn();
+        if (columnIndex > -1) {
+          currentRow.set(columnIndex, column);
+        } else {
+          currentRow.add(column);
+        }
+        colspan--;
+      }
+    }
+
+    private void newColumnWithRowSpan(Table.Column column) {
+      int rowspan = getRowSpan(column.node());
+      int rowStart = rows.indexOf(currentRow);
+      int rowEnd = rowStart + rowspan;
+      int columnIndex = rows.get(rowStart).indexOfVacantColumn();
+      if (columnIndex == -1) {
+        columnIndex = rows.get(rowStart).size();
+      }
+      for (int row = rowStart; row < rowEnd; ++row) {
+        if (row == rows.size()) {
+          rows.add(new RowBuilder());
+        }
+        if (columnIndex < rows.get(row).size()) {
+          rows.get(row).set(columnIndex, column);
+        } else {
+          for (int i = rows.get(row).size(); i < columnIndex; ++i) {
+            rows.get(row).add(null);
+          }
+          rows.get(row).add(column);
+        }
+      }
+    }
+
+    private void newColumnWithRowColSpan(Table.Column column) {
+      int rowspan = getRowSpan(column.node());
+      int rowStart = rows.indexOf(currentRow);
+      int rowEnd = rowStart + rowspan;
+      int colspan = getColSpan(column.node());
+      int columnStart = rows.get(rowStart).indexOfVacantColumn();
+      if (columnStart == -1) {
+        columnStart = rows.get(rowStart).size();
+      }
+      int columnEnd = columnStart + colspan;
+      for (int row = rowStart; row < rowEnd; ++row) {
+        if (row == rows.size()) {
+          rows.add(new RowBuilder());
+        }
+        for (int col = columnStart; col < columnEnd; ++col) {
+          if (col < rows.get(row).size()) {
+            rows.get(row).set(col, column);
+          } else {
+            for (int i = rows.get(row).size(); i < col; ++i) {
+              rows.get(row).add(null);
```

Comment:
I would avoid using `null` as value in collection, you can perhaps create an instance of `Column` class to represent this value